### PR TITLE
feat: Pluggable auth support

### PR DESCRIPTION
**Pull request by renkelvin**
Adding pluggable auth support.

### CODE REVIEW

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:305]: Let's reduce nesting in this block and remove the elif statements since we're raising errors.

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py:159]: Also validate that it is between 0-120s.

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:202]: You need to parse the URL for the email, and set the email here not URL.

Comment (lsirac) [reply google/auth/pluggable.py:202]: There is a service_account_email() function in the base class that returns it.

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:183]: I think ignoring issues when trying to read from the file hides bugs and maybe we should just throw an error. WDYT?

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:306]: If success=false, the error will look like this:

{
  "version": 1,
  "success": false,
  "code": "401",
  "message": "Permission denied. Caller not authorized"
}
Both 'code' and 'message' are required in this case. You need to build the error message using these fields.

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:309]: This is the first thing we should check

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:315]: Do we have any errors that have a code and a message?

For instance, if expired, we can return {INVALID_RESPONSE, "The executable response is expired.")

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:181]: Let's add a comment that explains that if the cached response is expired, _parse_subject_token will raise an error which will be ignored and we will call the executable again.

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py:17]: Also, you need to mention the executable specification, or point them to the README.

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:187]: Nit: A few comments are missing periods :-)

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:211]: Does this raise an error if the executable times out or is interrupted? We need to throw a custom error if it doesn't run within the specified timeout

**Review by lsirac — COMMENTED** [review]
First pass. I didn't review the tests. Thanks Chuan!

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:211]: The error is returned in result and we just raise it afterwards. What does the custom error look like?

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:315]: I don't see any errors with error code though. They're pure messages.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:183]: The output file is more of a cache here. So if the file is broken, we just invalidate the cache and retrieve the token again. Throwing an error requires the developer to manually correct the output file, which might be hard.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:218]: Yeah, it seems to be global in python.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:69]: Make sense. Added a check to make sure it's none.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:17]: Yeah, I duplicated the comments from external account credentials and added some comments for pluggable auth.
Make sense, added an example credential_source. Which README are you referring to?

**Comment by renkelvin** [comment]
Hmm, I can't push to the pluggable branch. I may need to create a new pr from my fork.

**Comment by renkelvin** [comment]
Hey @arithmetic1728, mind giving me the temporary write access to this repo? It should make the review much easier.

**Comment by arithmetic1728** [comment]
@renkelvin you need to join https://github.com/googleapis first, then I can give you temporary write access

**Comment by renkelvin** [comment]
> @renkelvin you need to join https://github.com/googleapis first, then I can give you temporary write access

Done joining https://github.com/googleapis

**Comment by arithmetic1728** [comment]
@renkelvin I just gave you the write access, let me know if it doesn't work.

**Comment by renkelvin** [comment]
> @renkelvin I just gave you the write access, let me know if it doesn't work.

It works, thanks!

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py]: Ok. In Java, I defined a new PluggableAuthException. Maybe we can do something similar here?

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py]: Looks like we document stuff here:

https://github.com/googleapis/google-auth-library-python/blob/main/docs/user-guide.rst

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py:180]: If the file is corrupted, or acl'd badly, or something else, then we are hiding this from the developer and it is not behaving as intended. I think throwing an error here instead ensures that their configuration is properly set up and everything is WAI.

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py:208]: I defined a new PluggableAuthException that has an error code and error description.

e.g.

```
...
throw new PluggableAuthException(
    "INTERRUPTED", String.format("The executable was interrupted: %s.", e));
```

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:155]: Need to do this check in milliseconds.

Also I said [0,120] but it is actually [5,120].

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:308]: why is this repeated?

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:129]: nit: missing space
 
and maybe change to

"Workforce Pools are not yet supported with Pluggable Auth."

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py]: Mentioned in the spec in go/pluggable-auth-3pi

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:180]: Make sense.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py]: Looks better, let me try add something similar.

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py]: You can actually pass the env to subprocess.run and it won't inherit from the current process:

```
env = os.environ.copy()
// set mappings on the copy
...
subprocess.run(env, ...)
```

Then you don't need to reset.

**Comment by renkelvin** [comment]
@arithmetic1728 Any ideas how can I add `pytest_subprocess` as a dev dependency just for testing purpose? Thanks! https://source.cloud.google.com/results/invocations/e6b81a67-add0-4bfa-9089-fe92b4cc7007/targets/github%2Fgoogle-auth-library-python/tests

**Comment by arithmetic1728** [comment]
@renkelvin I think you can add it here: https://github.com/googleapis/google-auth-library-python/blob/main/testing/requirements.txt

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py]: There are multiple cases that we should rerun the executable. For example: the file is not found, the token is expired, etc.. So we can't just throw an exception here.
How about we rerun the executable, but print in the console to inform the developer that the cached output file is not valid?

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py]: We throw 2 types of error here: ValueError(https://docs.python.org/3/tutorial/errors.html#raising-exceptions) and google.auth.exceptions(https://google-auth.readthedocs.io/en/stable/reference/google.auth.exceptions.html#google.auth.exceptions.RefreshError). Both of them take only a string to construct.

I tried to define an error class, however I think it makes the error construction more complicated and less clear. It also makes it hard for the system to read the string literal of the error class. WDYT?

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py]: Done. Good point.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:271]: I'm confused here. What's the difference between "validate the response itself" and "check that the response is valid"? Can you redirect me to your java code?

File: tests/test_pluggable.py
Comment (renkelvin) [reply tests/test_pluggable.py:51]: 1. Added: unsuccessful response with error code and error message; unsupported token type;
Existing: unsupported response version; token expired;
For 2 & 3, due to the limitation of `pytest_subprocess` package, I didn't find a way to check the env variables of the subprocess, or set a running time for the subprocess. `FILE_NOT_FOUND` exception is covered, where the exception is ignored and we rerun the executable.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:180]: Done.

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py:271]: There are two things: 
1) Validating the executable response format -> is there a "success" field? is there a "version" field? etc. if not, we can throw an error saying this field is missing
2) Validating the response itself -> is the "version" field set to 1?, etc.

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py]: Let's keep it like this then if this is how it currently works in the lib.

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py:180]: Seems like there is missing a change here? Still just ignoring any issues with the output file?

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py:205]: If the executable returns a bad response, what does the error look like right now?

File: google/auth/_default.py
Comment (lsirac) [inline google/auth/_default.py:320]: Seems like you forgot to update ADC?

File: google/auth/pluggable.py
Comment (lsirac) [inline google/auth/pluggable.py:205]: I think this will result in this process hanging if the STDOUT buffer is filled. Can you test (go/pluggable-auth-sdk-testing - last test case) and provide a workaround?

Comment (renkelvin) [reply google/auth/pluggable.py:205]: Wow, good case. Let me test it out.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:205]: See the lines below. The error will be something like:
"Executable exited with non-zero return code 1. Error: DETAILED ERROR MESSAGE FROM STDOUT"

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:180]: If there is an issue with the file itself (not the content) like file not exists, the error is ignored and we generate the file then.
If the there is an issue with the content, the error is raised.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:209]: Confirmed that it won't hang but error out with invalid response.

**Comment by renkelvin** [comment]
Hey @arithmetic1728, it seems the build for Python 3.7 passed but Kokro build failed.
Is the "Kokoro" build for 2.x versions? Received an error of "ModuleNotFoundError: No module named 'pytest_subprocess'". Note that `pytest_subprocess` is only available for Python 3.x. Any ideas? Thanks!

**Comment by arithmetic1728** [comment]
@renkelvin It seems both python 3.6 and 2.7 unit tests failed with this error. `google-auth` supports python 2.7 and >= 3.6. https://github.com/googleapis/google-auth-library-python/blob/main/setup.py#L65

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py:280]: Looks like you added it. Resolving.

**Review by lsirac — APPROVED** [review]
LGTM for implementation details. Thanks Chuan!

**Comment by renkelvin** [comment]
@arithmetic1728 I thought 2.x is no longer supported, according to the README file https://github.com/googleapis/google-auth-library-python/blob/main/README.rst
Is it possible we support this feature somehow only for 3.6+? Since it requires the timeout support of `subprocess` module which is only available in python 3.4+.

**Comment by arithmetic1728** [comment]
@renkelvin We don't need to support new features for 2.7, but the code has to be 2.7 compatible. gcloud CLI uses google-auth and it still supports 2.7, so if we use the 3.x only API in google-auth, gcloud will break.

I guess you can do the following:
```
import sys

def refresh(..):
    if sys.version_info < (3, 0):
        raise RefreshError("pluggable auth is only supported for python 3.6+")
```

If you want to support it for 2.7, then you can implement the timeout on your own.

File: google/auth/pluggable.py
Comment (sai-sunder-s) [inline google/auth/pluggable.py:281]: I am confused about the purpose of the version. This is not  a version for the credential_source format, but its a version for the output of the executable?

**Comment by renkelvin** [comment]
> @renkelvin We don't need to support new features for 2.7, but the code has to be 2.7 compatible. gcloud CLI uses google-auth and it still supports 2.7, so if we use the 3.x only API in google-auth, gcloud will break.
> 
> I guess you can do the following:
> 
> ```
> import sys
> 
> def refresh(..):
>     if sys.version_info < (3, 0):
>         raise RefreshError("pluggable auth is only supported for python 3.6+")
> ```

@arithmetic1728 Thanks! After some discussions with my teammates, we decide to support both 2.7 and 3.6. The time timeout support will be added in a separate pr though.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:281]: Yes, the output of the executable is expected to comply with the spec(versioned) we defined.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:196]: They are used by the customer provided plugged executable.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:283]: We might stop supporting old versions in the future, while EXECUTABLE_SUPPORTED_MIN_VERSION is not in scope yet and it can be added later I think. @lsirac any thoughts?

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:226]: Yes, the file will be used as cache.

File: google/auth/pluggable.py
Comment (renkelvin) [reply google/auth/pluggable.py:220]: See line 213, the output of stderr is redirected to stdout.

File: google/auth/pluggable.py
Comment (lsirac) [reply google/auth/pluggable.py:283]: I don't think it makes sense to have a min version. We generally don't make backwards incompatible changes. The use case here is if we update the response specification and try to use an old library. If needed it can be added in the future.

**Comment by arithmetic1728** [comment]
You can run the following easily-to-fail kokoro tests locally before pushing to github. This will save a lot of time. (Don't worry about the kokoro system-3.7 failure - It is caused by expired creds. I will update the creds once your PR is ready to merge).

lint: `python -m nox -s lint`
If lint fails, run `python -m nox -s blacken` to format the files, then re-run lint. If it can pass, don't forget to add and commit the changes.

cover: `python -m nox -s cover`

**Comment by renkelvin** [comment]
> You can run the following easily-to-fail kokoro tests locally before pushing to github. This will save a lot of time. (Don't worry about the kokoro system-3.7 failure - It is caused by expired creds. I will update the creds once your PR is ready to merge).
> 
> lint: `python -m nox -s lint` If lint fails, run `python -m nox -s blacken` to format the files, then re-run lint. If it can pass, don't forget to add and commit the changes.
> 
> cover: `python -m nox -s cover`

Thanks @arithmetic1728! The CI passes now.

**Review by arithmetic1728 — APPROVED** [review]

**Comment by arithmetic1728** [comment]
LGTM 

@sai-sunder-s would you like to take another look?
@renkelvin I will fix the system test in a separate PR when we are ready to check in.

**Review by sai-sunder-s — APPROVED** [review]

**Pull request merged**
